perf-cluster-operation/pkg/cluster.go

```diff
@@ -0,0 +1,242 @@
+	if cluster.Status == statusProvisioning || cluster.Status == statusStopping {
+		log.Printf("cluster %q is being handled by other jobs, skip it", cluster.Name)
+		return nil
+	}
+
+	config, configExists := clusterConfigs[cluster.Name]
+	// if retainIfUnchanged is set to true, and the cluster config does not change, do nothing
+	// TODO(chizhg): also check the addons config
+	if retainIfUnchanged && configExists &&
+		cluster.CurrentNodeCount == config.NodeCount && cluster.Location == config.Location {
+		log.Printf("cluster config is unchanged for %q, skip it", cluster.Name)
+		return nil
+	}
+
+	if err := gc.deleteClusterWithRetries(gcpProject, cluster); err != nil {
+		return fmt.Errorf("failed deleting cluster %q in %q: %v", cluster.Name, cluster.Location, err)
+	}
+	if configExists {
+		return gc.createClusterWithRetries(gcpProject, cluster.Name, config)
+	}
+	return nil
+}
+
+// listClustersForRepo will list all the clusters under the gcpProject that belong to the given repo.
+func (gc *gkeClient) listClustersForRepo(gcpProject, repo string) ([]container.Cluster, error) {
+	allClusters, err := gc.ops.ListClustersInProject(gcpProject)
+	if err != nil {
+		return nil, fmt.Errorf("failed listing clusters in project %q: %v", gcpProject, err)
+	}
+
+	clusters := make([]container.Cluster, 0)
+	for _, cluster := range allClusters {
+		if clusterBelongsToRepo(cluster.Name, repo) {
+			clusters = append(clusters, *cluster)
+		}
+	}
+	return clusters, nil
+}
+
+// deleteClusterWithRetries will delete the given cluster,
+// and retry for a maximum of retryTimes if there is an error.
+// TODO(chizhg): maybe move it to clustermanager library.
+func (gc *gkeClient) deleteClusterWithRetries(gcpProject string, cluster container.Cluster) error {
+	region, zone := gke.RegionZoneFromLoc(cluster.Location)
+	var err error
+	for i := 0; i < retryTimes; i++ {
+		if err = gc.ops.DeleteCluster(gcpProject, region, zone, cluster.Name); err == nil {
+			break
+		}
+	}
+	if err != nil {
+		return fmt.Errorf(
+			"failed deleting cluster %q in %q after retrying %d times: %v",
+			cluster.Name, gke.GetClusterLocation(region, zone), retryTimes, err)
+	}
+
+	return nil
+}
+
+// createClusterWithRetries will create a new cluster with the given config,
+// and retry for a maximum of retryTimes if there is an error.
+// TODO(chizhg): maybe move it to clustermanager library.
+func (gc *gkeClient) createClusterWithRetries(gcpProject, name string, config ClusterConfig) error {
+	req := &gke.Request{
+		ClusterName: name,
+		MinNodes:    config.NodeCount,
+		MaxNodes:    config.NodeCount,
+		NodeType:    config.NodeType,
+		Addons:      strings.Split(config.Addons, ","),
+	}
+	creq, err := gke.NewCreateClusterRequest(req)
+	if err != nil {
+		return fmt.Errorf("cannot create cluster with request %v: %v", req, err)
+	}
+
+	region, zone := gke.RegionZoneFromLoc(config.Location)
+	for i := 0; i < retryTimes; i++ {
```

Comment:
if creation failed after it started, then retry for sure will fail since the old cluster still exists

Reply:
Yes that's true.. So we probably should delete the old cluster and then retry, based on the error type? Does this problem also apply to cluster setup for e2e tests?

Reply:
Added the deletion logic.
I think `Wait` won't finish when the cluster is still in `PROVISIONING` state, unless something really bad happens, then some operations will fail anyway.